### DISCUSSION TITLE
MATLAB Controller using JSBSim S-Function

### DISCUSSION THREAD

**Discussion by hjorthchristian** [JSBSim-Team/jsbsim#992] [Q&A]
I am interested to know if there is a repository or if anyone could provide any guidance on how to control an aircraft with the help of the S-Function provided by JSBSim in MATLAB. However, I am new to this world and I am trying to learn. I have tested their tutorial code with the 737 example and it works and I have seen that there are some necessary parameters for the simulation to run, however I find it difficult to make the function work to any arbitrary aircraft, maybe it has to do with the IO xml file, but it seems that I must have a specific number of inputs in order for it to work (8 inputs?). To sum it up, I am interested in making a controller for an aircraft, however I am having some difficulties. Can I extract a model of an aircraft from JSBSim and use it in matlab? How is the standard way to use the JSBSim S-function for controlling the aircraft? Any help at all is very welcome! :)

**Comment by bcoconni** [comment]
Hi @hjorthchristian,

At least theoretically any JSBSim model can be used with the S-Function.
@agodemar or @noah-curran may provide you some help as they are advanced users of the Matlab S-Function.

**Comment by noah-curran** [comment]
Yes, this should be possible.
You'll need to edit the IO configuration file to have all of the necessary input required by your custom aircraft.
You'll probably need to take a look at the aircraft's XML and edit the IO config file to include the exact name for the properties used there (as well as their exact location in the property tree). This will require some effort from you since we cannot guess what the properties are named in the flight control scripts.

**Reply by hjorthchristian** [reply]
Thanks for the answer! I've made the function work with some other aircrafts, however is it always necessary to give 8 inputs? Is it also possible to extract EoM from JSBSim or how should someone ensure that they have a model that is correspondant in MATLAB and JSBSim? I ask this because I would like to control the inputs through MATLAB, however I'm wondering how I can extract the necessary information from JSBSim to make them match.

**Reply by agodemar** [reply]
@hjorthchristian looking into the code is usually the first option you have. It should be possible to define the outputs of the S-Function and use them accordingly in Simulink.

**Reply by bcoconni** [reply]
> I've made the function work with some other aircraft, however is it always necessary to give 8 inputs? 

According to the code below, it does not seem that the S-Function is enforcing a particular number of input parameters (or output parameters by the way). @noah-curran authored that code so he will correct me if I'm wrong but it seems that you can use as many or as few parameters as you need.

https://github.com/JSBSim-Team/jsbsim/blob/48e2558a7620918212e5de117f19b97f7b587820/matlab/JSBSim_SFunction.cpp#L305-L318
>  Is it also possible to extract EoM from JSBSim

JSBSim exports a fair amount of properties about the position, attitude, velocities and accelerations of the vehicle so you can check exactly what the EOM are computing at any time. Hopefully you should be able to access the data you need.

https://github.com/JSBSim-Team/jsbsim/blob/48e2558a7620918212e5de117f19b97f7b587820/src/models/FGPropagate.cpp#L825-L886
https://github.com/JSBSim-Team/jsbsim/blob/48e2558a7620918212e5de117f19b97f7b587820/src/models/FGAccelerations.cpp#L348-L362

**Reply by noah-curran** [reply]
Yes, as many or as few input and output properties that are wanted. You will just need to change them in the IO config file. You can group the output up into as many groups as wanted as well. (The input all need to be in the same group.)
@hjorthchristian if you define additional properties in your aircraft XML file, using something like fcs_function, you can also extract those as well.

**Reply by hjorthchristian** [reply]
Thank you all very much for the reply, It has helped me understand better the S-Function! I Have one question regaridng making a controller. For example when you set the ap/altitude_setpoint with the C172, and then the autopilot takes the airplane to the deisred altitude. Can you you reproduce it but with MATLAB instead of the "inbuilt" C172 AP? Would you then make it for example fly at an altitude with no altitude_setpoints in the mission script but rather have a trigger in MATLAB according to some condition?

**Reply by noah-curran** [reply]
Yes. This exact purpose was the reason why I made the S-Function's IO configurable.

**Comment by seanmcleod** [comment]
Yes, you are not forced to use any existing auto-pilot systems. You can implement your own controller completely in Matlab/Simulink. You obviously need to ensure that in terms of IO config file that you're including all the required states that your controller requires.

**Reply by hjorthchristian** [reply]
Perfect! that is great to hear! Is it possible to run a script without feeding input from MATLAB, bascially do the same thiing as JSBSim would in its normal mission script. I ask this for more of a comparison as to have for example a mission script in JSBSim that is just cruising, and then use the same script in MATLAB without any input to see if JSBSim and MATLAB's S-Function match their outputs, and then start editing from there.

**Comment by seanmcleod** [comment]
Sure, create an IO file which has no inputs from Matlab/Simulink. That way the only inputs during the run will come from the script file.
